From 0a0ddcf28fd6a8d80fb677bcc5ab866441ed1caf Mon Sep 17 00:00:00 2001
From: Tim Nordell <tnordell@airgain.com>
Date: Thu, 12 Mar 2026 11:59:17 -0500
Subject: [PATCH 7/7] jail: Drop privileges after setting up the jail

When we're idling for the ubus relay, we should drop as many privileges
as possible in order to sandbox ourselves.  We do three things:

  - Switch our root into /var/run/ubus
    - This limits what parts of the filesystem is visible to us to just
      being able to interact with the ubus.sock file
  - Drop all capabilities
    - We don't need any capabilities once everything is setup
  - Setup seccomp to just the system calls we require

This should effectively sandbox ujail at this point, and permit it to
still be "root" so that it has full access to ubus.

(Although, having access to ubus does give it technically the ability to
breakout further, but since this is limited to ubus filtering at this point,
hopefully this is sufficient.)
---
 jail/jail.c | 69 +++++++++++++++++++++++++++++++++++++++++++++++++++++
 jail/ubus.c |  4 ++--
 2 files changed, 71 insertions(+), 2 deletions(-)

diff --git a/jail/jail.c b/jail/jail.c
index 6d4eae3..ecb0231 100644
--- a/jail/jail.c
+++ b/jail/jail.c
@@ -163,6 +163,43 @@ static struct {
 	bool ubus_rpc_relay;
 } opts;
 
+const char jail_seccomp[] =
+"{"
+"       \"defaultAction\": \"SCMP_ACT_ERRNO\","
+"       \"defaultErrnoRet\": 1,"
+"       \"syscalls\": ["
+"               {"
+"                       \"names\": ["
+"                               \"accept\","
+"                               \"close\","
+"                               \"connect\","
+"                               \"epoll_ctl\","
+"                               \"epoll_pwait\","
+"                               \"exit_group\","
+"                               \"fcntl\","
+"                               \"getpid\","
+"                               \"getsockopt\","
+"                               \"mmap\","
+"                               \"munmap\","
+"                               \"pidfd_send_signal\","
+"                               \"ppoll\","
+"                               \"read\","
+"                               \"recvmsg\","
+"                               \"rt_sigaction\","
+"                               \"rt_sigprocmask\","
+"                               \"rt_sigreturn\","
+"                               \"sendmsg\","
+"                               \"sendto\","
+"                               \"socket\","
+"                               \"wait4\","
+"                               \"write\","
+"                               \"writev\""
+"                       ],"
+"                       \"action\": \"SCMP_ACT_ALLOW\""
+"               }"
+"       ]"
+"}";
+
 static struct blob_buf ocibuf;
 
 extern int pivot_root(const char *new_root, const char *put_old);
@@ -3204,8 +3241,40 @@ static void pipe_send_start_container(struct uloop_timeout *t)
 	run_hooks(opts.hooks.poststart, post_poststart);
 }
 
+static void drop_privileges(void)
+{
+	struct blob_buf b = { 0 };
+	struct sock_fprog *prog = NULL;
+	struct jail_capset capset = {0};
+
+	/* Ensure syslog is open before we chroot;
+	 * notably, this does not survive logd restarting after being in the
+	 * chroot(), but we'll live with that.
+	 */
+	openlog(NULL, LOG_PID | LOG_CONS | LOG_NDELAY, LOG_DAEMON);
+
+	/* Switch root into ubus's socket folder */
+	chroot("/var/run/ubus");
+
+	/* Drop all capabilities */
+	capset.apply = 1;
+	applyOCIcapabilities(capset, 0);
+
+	/* Setup seccomp */
+	blob_buf_init(&b, 0);
+	if (!blobmsg_add_json_from_string(&b, jail_seccomp)) {
+		blob_buf_free(&b);
+		ERROR("can't process jail_seccomp\n");
+		free_and_exit(-1);
+	}
+	prog = parseOCIlinuxseccomp(b.head);
+	applyOCIlinuxseccomp(prog);
+	blob_buf_free(&b);
+}
+
 static void post_poststart(void)
 {
+	drop_privileges();
 	uloop_run(); /* idle here while jail is running */
 	if (jail_running) {
 		DEBUG("uloop interrupted, killing jail process\n");
diff --git a/jail/ubus.c b/jail/ubus.c
index 271e409..86d5ed6 100644
--- a/jail/ubus.c
+++ b/jail/ubus.c
@@ -672,7 +672,7 @@ static void ubus_relay_server_cb(struct uloop_fd *fd, unsigned int events)
 		return;
 	}
 
-	ubus_fd = usock(USOCK_UNIX, UBUS_UNIX_SOCKET, NULL);
+	ubus_fd = usock(USOCK_UNIX, "/ubus.sock", NULL);
 	if (ubus_fd < 0) {
 		close(jail_fd);
 		return;
@@ -707,7 +707,7 @@ static void ubus_relay_ubus_connected(struct ubus_relay_server *server) {
 
 static void ubus_relay_ubus_timeout(struct uloop_timeout *ctx) {
 	struct ubus_relay_server *server = container_of(ctx, struct ubus_relay_server, ubus_reconnect_tmo);
-	if (ubus_connect_ctx(&server->ubus, NULL) == 0) {
+	if (ubus_connect_ctx(&server->ubus, "/ubus.sock") == 0) {
 		ubus_relay_ubus_connected(server);
 	} else {
 		uloop_timeout_set(&server->ubus_reconnect_tmo, 1000);
-- 
2.53.0

